### PR TITLE
Make WebRequest use HttpClient and proper async/await

### PR DESCRIPTION
**Pull request by smoogipoo**
Optional:
- [x] https://github.com/ppy/osu-framework/pull/1090

Found [this](https://github.com/dotnet/corefx/issues/11885#issuecomment-248915149) comment during my quest to fix the WebRequest issue on .NET Core. Furthermore [this](https://github.com/dotnet/corefx/issues/7037) thread basically states that:
1. Anything to do with WebRequest is deprecated.
2. `HttpClient` handles the proxy stuff much better.

As far as I can see, `HttpClient` is the only way forward for .NET Core compatibility.

This branch serves two main purposes:
1. Making `WebRequest` use `HttpClient` for requests.
2. Utilising async/await methods to make the code neater.

The following is _not_ implemented:
1. Certificate validation. This needs to be re-implemented at an osu.Game level - it cannot enforce `*ppy.sh` certificates.
2. IPv4 fallback. I don't know if this is needed and will need further testing to make sure .NET (and core) doesn't just handle this for us.
3. Enforcing maximum 12 connections. This isn't available in the .NET Framework HttpClient (`MaxConnectionsPerServer`) and has been commented out. This is available in System.Net.Http/4.3.3.

The following has _not_ been tested:
1. `POST` file requests. Implementation looks 1:1 as far as I can see though.

### CODE REVIEW

File: osu.Framework/IO/Network/WebRequest.cs [hunk 1]
```diff
@@ -124,20 +125,29 @@ public string Url
...
         static WebRequest()
         {
-            //set some sane defaults for ServicePoints
-            ServicePointManager.Expect100Continue = false;
-            ServicePointManager.DefaultConnectionLimit = 12;
-            ServicePointManager.CheckCertificateRevocationList = false;
-            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls;
+            var handler = new HttpClientHandler
+            {
+                AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
```
Comment (peppy) [inline osu.Framework/IO/Network/WebRequest.cs]: this is the default and can be omitted

Comment (smoogipoo) [reply osu.Framework/IO/Network/WebRequest.cs]: Where are you seeing this? https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpClientHandler.Windows.cs#L39 + https://github.com/dotnet/corefx/blob/93ee4ba40c82d5aca978447cb3e14c4ef7e7fd53/src/Common/src/System/Net/Http/HttpHandlerDefaults.cs#L15 doesn't look like defaults to me.

Comment (smoogipoo) [reply osu.Framework/IO/Network/WebRequest.cs]: As discussed, it is none by default as of https://github.com/dotnet/corefx/commit/eb21a724bff127d3ab0c34240234ef93ac1a94fb#diff-bfbd00735543382ef8b5e517ed3acde6 and the aforementioned linked code.

**Comment by peppy** [comment]
Any reason for not also providing an non-async `Perform` method? I *think* it's still good practice to do this.

**Comment by smoogipoo** [comment]
Yeah I think you're right, will add back a non-async method.

File: osu.Framework/IO/Network/WebRequest.cs [hunk 2]
```diff
@@ -285,126 +206,110 @@ public byte[] ResponseData
...
-                        request = CreateWebRequest();
-                        request.Method = @"POST";
+                    switch (Method)
+                    {
+                        case HttpMethod.GET:
+                            Debug.Assert(Files.Count == 0);
 
-                        if (Parameters.Count + Files.Count == 0)
-                        {
-                            rawContent?.WriteTo(requestBody);
```
Comment (peppy) [inline osu.Framework/IO/Network/WebRequest.cs]: this was removed and not added back. `AddRaw` is completely non-functional atm.

**Comment by peppy** [comment]
Still planning on adding back non-async perform?

**Comment by smoogipoo** [comment]
Yup, I completely forgot about it.

File: osu.Framework.Tests/IO/TestWebRequest.cs [hunk 3]
```diff
@@ -18,7 +18,7 @@ public class TestWebRequest
 
         [TestCase("http")]
         [TestCase("https")]
-        public void TestValidGet(string protocol)
+        public void TestValidGetAsync(string protocol)
```
Comment (peppy) [inline osu.Framework.Tests/IO/TestWebRequest.cs]: this method's renamed, but all the other tests also only test async.

i think it'd be good to have a test of non-async to make sure exceptions are correctly being thrown outwards from `Perform`

File: osu.Framework/IO/Network/WebRequest.cs [hunk 4]
```diff
@@ -285,202 +206,174 @@ public byte[] ResponseData
...
+                                {
+                                    var byteContent = new ByteArrayContent(p.Value);
+                                    byteContent.Headers.Add("Content-Type", "application/octet-stream");
+                                    formData.Add(byteContent, p.Key, p.Key);
+                                }
+
+                                contentLength = formData.ReadAsByteArrayAsync().Result.Length;
+
+                                requestStream = new LengthTrackingStream(formData.ReadAsStreamAsync().Result);
+                                requestStream.BytesRead.ValueChanged += v =>
```
Comment (peppy) [inline osu.Framework/IO/Network/WebRequest.cs]: this whole raw section is some sloppy coding (and doesn't have unit tests to cover it).

File: osu.Framework/IO/Network/WebRequest.cs [hunk 5]
```diff
@@ -244,7 +244,10 @@ public Task PerformAsync()
                         default:
                             throw new InvalidOperationException($"HTTP method {Method} is currently not supported");
                         case HttpMethod.GET:
-                            Debug.Assert(Files.Count == 0);
+                            if (Parameters.Count > 0)
+                                throw new InvalidOperationException($"Cannot use {nameof(AddParameter)} in a GET request. Please set the {nameof(Method)} to POST.");
```
Comment (smoogipoo) [inline osu.Framework/IO/Network/WebRequest.cs]: We can have Parameters with GET requests, just not files or raw content.

File: osu.Framework/IO/Network/WebRequest.cs [hunk 6]
```diff
@@ -213,7 +212,12 @@ public byte[] ResponseData
         private CancellationTokenSource cancellationToken;
         private LengthTrackingStream requestStream;
         private HttpResponseMessage response;
-        private long contentLength;
+
+        private long contentLength => requestStream?.Length ?? 0;
+
+        private const string form_boundary = "-----------------------------28947758029299";
+
+        private const string form_content_type = "multipart/form-data; boundary=" + form_boundary;
```
Comment (smoogipoo) [inline osu.Framework/IO/Network/WebRequest.cs]: Does string interpolation not work on consts?

Comment (peppy) [reply osu.Framework/IO/Network/WebRequest.cs]: doesn't

**Pull request merged**